### PR TITLE
Remove use of ndarrays in Fortran array argument wrapping

### LINKED ISSUE #1962: Remove use of ndarrays when passing arrays as arguments to Fortran functions
### Describe the feature
As a first step towards implementing #1961 we can remove the use of ndarrays when passing arrays as arguments to Fortran functions. Currently ndarrays are used as an intermediate step in the cwrapper when passing from Python to C. 

### Current Code
```c
static PyObject* bind_c_array_argument_wrapper(PyObject* self, PyObject* args, PyObject* kwargs)
{
    PyObject* x_obj;
    t_ndarray x = {.shape = NULL};
    void* bound_x;
    int64_t bound_x_shape_1;
    int64_t bound_x_shape_2;
    int64_t bound_x_stride_1;
    int64_t bound_x_stride_2;
    static char *kwlist[] = {
        "x",
        NULL
    }; 
    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O", kwlist, &x_obj))
    {   
        return NULL;
    }
    if (pyarray_check(x_obj, NPY_LONG, INT64_C(2), NPY_ARRAY_C_CONTIGUOUS))
    {   
        x = pyarray_to_ndarray(x_obj);
    }
    else
    {
        PyErr_SetString(PyExc_TypeError, "Expected an argument of type numpy.ndarray[numpy.int64] for argument x");
        return NULL;
    }
    bound_x = nd_data(&x);
    bound_x_shape_1 = nd_ndim(&x, INT64_C(0));
    bound_x_shape_2 = nd_ndim(&x, INT64_C(1));
    bound_x_stride_1 = nd_nstep_C(&x, INT64_C(0));
    bound_x_stride_2 = nd_nstep_C(&x, INT64_C(1));
    bind_c_array_argument(bound_x, bound_x_shape_1, bound_x_shape_2, bound_x_stride_1, bound_x_stride_2);
    free_pointer(&x);
    Py_INCREF(Py_None);
    return Py_None;
}
```

### Proposed Solution
```c
static PyObject* bind_c_array_argument_wrapper(PyObject* self, PyObject* args, PyObject* kwargs)
{
    PyObject* x_obj;
    void* x_data;
    int64_t x_shape[2];
    int64_t x_stride[2];
    static char *kwlist[] = {
        "x",
        NULL
    }; 
    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O", kwlist, &x_obj))
    {   
        return NULL;
    }
    if (pyarray_check(x_obj, NPY_LONG, INT64_C(2), NPY_ARRAY_C_CONTIGUOUS))
    {
        x_data = PyArray_DATA(x_obj);
        get_strides_and_size_from_numpy_array(x_obj, &x_shape, &x_stride, false);
    }
    else
    {
        PyErr_SetString(PyExc_TypeError, "Expected an argument of type numpy.ndarray[numpy.int64] for argument x");
        return NULL;
    }
    bind_c_array_argument(x_data, x_shape[0], x_shape[1], x_stride[0], x_stride[1]);
    Py_INCREF(Py_None);
    return Py_None;
}
```

### PR DESCRIPTION
**Pull request by EmilyBourne**
Remove the use of `ndarray_t` in the C wrapper of a function translated to Fortran. Fixes #1962 , making it easier to remove `ndarray_t` later and reducing complexity. 

The functions `_wrap_FunctionDefArgument` and `_wrap_BindCFunctionDefArgument` are modified to return a dictionary with the useful information that they calculate. In addition to the body returned previously the function arguments are additionally returned. This makes it easier to pass multiple variables to the function and simplifies `_wrap_FunctionDef` slightly 

**Commit Summary**
- Add the function `get_strides_and_shape_from_numpy_array` to calculate the strides and shape of a NumPy array
- Use `var` to avoid the repetition of `expr.variable` in `CCodePrinter_print_Declare`
- Add code to declare a C stack array
- Use consistent naming in `c_to_python_wrapper.py`. The original name is now always the name that is used to recognise a variable in the scope
- Add `_get_array_parts` to get the data, shape, and strides of an array
- Modify `_wrap_FunctionArgument` and `_wrap_BindCFunctionArgument` to return dictionaries for readability. The arguments passed to the function are also returned so operations such as adding `ObjectAddress` can be done in the relevant function instead of being added a posteriori in `_wrap_FunctionDef`

### CODE REVIEW

File: pyccel/codegen/printing/ccode.py [hunk 1]
```diff
@@ -1379,14 +1383,22 @@ def _print_FuncAddressDeclare(self, expr):
...
 
-        if expr.variable.is_stack_array:
-            preface, init = self._init_stack_array(expr.variable,)
+        if isinstance(var.class_type, CStackArray):
+            preface = ''
+            if isinstance(var.shape[0], LiteralInteger):
+                init = f'[{var.shape[0]}]'
+            else:
+                declaration_type += '*'
+                init = ''
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:1399]: /bot accept
This code is non-critical but easy to specify. It is simpler to specify it than to raise a non-implemented error

**Comment by EmilyBourne** [comment]
PR is stuck due to:
https://stackoverflow.com/questions/78919927/how-to-debug-a-dylib-error-or-compiler-bug-in-a-python-c-api-function-wrapper

**Review by jalalium — APPROVED** [review]
Very good job!

File: pyccel/codegen/printing/ccode.py [hunk 2]
```diff
@@ -23,6 +23,8 @@
 from pyccel.ast.core      import SeparatorComment
 from pyccel.ast.core      import Module, AsName
 
+from pyccel.ast.c_concepts import CStackArray
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:27]: Oops. I thought these import files were alphabetised. Fixed: 280146c5e

File: pyccel/codegen/wrapper/c_to_python_wrapper.py [hunk 3]
```diff
@@ -1011,6 +1015,49 @@ def _get_class_destructor(self, del_function, cls_dtype, wrapper_scope):
 
         return function
 
+    def _get_array_parts(self, expr):
+        """
+        Get AST nodes describing the extraction of the data pointer, shape, and strides from a Python array object.
+
+        Get AST nodes describing the extraction of the data pointer, shape, and strides from a Python array object.
+        These noes as well as the new objects can then be packed into a structure or passed directly to a function
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/wrapper/c_to_python_wrapper.py]: Fixed e16873e2d

**Review by jalalium — APPROVED** [review]

**Pull request merged**